blob/fileblob/fileblob.go

```diff
@@ -492,22 +493,20 @@ func (w writer) Write(p []byte) (n int, err error) {
 			return 0, fmt.Errorf("updating md5 hash: %v", err)
 		}
 	}
-	return w.w.Write(p)
+	return w.f.Write(p)
 }
 
 func (w writer) Close() error {
-	// If the write was cancelled, delete the file.
-	if err := w.ctx.Err(); err != nil {
-		_ = os.Remove(w.path)
+	err := w.f.Close()
+	if err != nil {
 		return err
 	}
-	if err := setAttrs(w.path, w.attrs); err != nil {
-		return fmt.Errorf("write blob attributes: %v", err)
-	}
-	err := w.w.Close()
-	if err != nil {
+	// If the write was cancelled, delete the temp file.
+	if err := w.ctx.Err(); err != nil {
+		_ = os.Remove(w.f.Name())
```

Comment:
Can we `defer` this? Whenever we exit close we should all remove the temp file. On success, the `os.Remove` would return the "not exist" error which can be ignored.

Reply:
Good idea; done. I also changed it to write the attributes file first, so if that fails we can abort without having clobbered the original file (if any).